[Zhao Peiduo] iP

ZhaoPeiduo opened the pull request:
# Implementation for Duke

> This document presents the avaliable features and user commands. The following information may be used as a user guide for my implementaton of Duke. The original version of Duke is avaliable [here](https://github.com/nus-cs2103-AY2122S1/ip).

### Features :ballot_box_with_check:

- [x] Save and delete different types of tasks: todo, deadline, event;
- [x] Set a saved task to be done;
- [x] Autosave to and autoload user's data from a default directory;
- [x] Allows user to save/load data from a specified directory;
- [x] Search for task according to a specified key word.
- [x]  An in-app user guideline 

### Syntax List:
>Legend for datatype:  
>1. **bolded** String
>2. *italic* int
- todo <**description**>: create and save a todo task;
- deadline <**description**> /by <**time in format yyyy-mm-dd**>: create and save a deadline task;
- event <**description**> /at <**time in format yyyy-mm-dd**>: create and save an event task;
- save <**directory**>: save the current task list to the specified directory;
- load <**directory**>: load the task list from the specified directory;
- done <*number*>: mark the task with *number* index as done;
- delete <*number*>: delete the task with *number* index;
- deleteAll: delete all saved tasks in the task list;
- find <**keyword**>: find all tasks with description containing the **keyword**.
- help: shows the list of avaliable command

You may run the application by clicking the green triangle on the left of the main function in `Duke.class`.
```
public class Duke {
    ...
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

Commit:
Add Gradle support

ZhaoPeiduo committed:
Level-1 completed

ZhaoPeiduo committed:
Level-1 edited

ZhaoPeiduo committed:
Level-2 completed

ZhaoPeiduo committed:
Level-2 edited

ZhaoPeiduo committed:
Level-3 completed

ZhaoPeiduo committed:
Level-4 completed

ZhaoPeiduo committed:
Basic Unit test cases written

ZhaoPeiduo committed:
Level-5 completed, and testcase updated

ZhaoPeiduo committed:
Level-6 completed, and testcase updated

ZhaoPeiduo committed:
Complete Level-7: Automatically save and load users' input

ZhaoPeiduo committed:
Updated Level-6: Standardise indentation level and display format

ZhaoPeiduo committed:
Complete Level-8: Store deadline and event time in form of LocalDate

ZhaoPeiduo committed:
Merge branch 'branch-Level-7' with conflicts resolved

ZhaoPeiduo committed:
Merge branch 'branch-Level-8' ,update test cases

# Conflicts:
#	src/main/java/Duke.java

ZhaoPeiduo committed:
Complete Level-7 partially

ZhaoPeiduo committed:
Complete A-More OOP partially, wrongly indicate the task as Level-7 in the previous commit

ZhaoPeiduo committed:
Complete A-MoreOOP

ZhaoPeiduo committed:
Complete A-Packages: put all classes under the duke package

ZhaoPeiduo committed:
Complete A-JUnit

ZhaoPeiduo committed:
Break the Ui class into smaller subclasses, make Ui an interface and create methods for save/load to a specified file

ZhaoPeiduo committed:
Add JavaDoc to all methods and classes in Duke

ZhaoPeiduo committed:
Update test cases and add a few more

ZhaoPeiduo committed:
Change coding style to meet the coding standard

ZhaoPeiduo committed:
Complete Level9: Add find feature

ZhaoPeiduo committed:
Edit UI::formatMessage to use StringBuilder

ZhaoPeiduo committed:
- Change method names and add doc for constuctors to meet coding standard
- Add user guideline to Readme

ZhaoPeiduo committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ZhaoPeiduo committed:
Complete A-Gradle: build and run Duke with Gradle, set up gradle to run unit tests

ZhaoPeiduo committed:
Merge branch 'branch-A-Gradle'

ZhaoPeiduo committed:
Complete A-CheckStyle and used check style to solve coding standard issues
Implement the help command to allow user to access user guideline in Duke

ZhaoPeiduo committed:
Merge branch 'branchA-CheckStyle'

ZhaoPeiduo committed:
Finished setup of the sample GUI and changed original implementation to use the GUI

ZhaoPeiduo committed:
Change test file to fit the GUI and adjust javadoc by adding in @return tags

ZhaoPeiduo committed:
Merge branch 'branch-Level-10'

marcuspeh commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+ * Represents a specific type of Task with an additional deadline(ddl) field.
+ */
+public class Deadline extends Task {
+    private LocalDate ddl;
```
Is there any reason why this variable is ddl? Perhaps it will be better if it can be changed to a name that explains the name entiry.

ZhaoPeiduo replied:
Thanks for pointing this out. I have changed the variable name to deadline in the latest commit!

marcuspeh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,113 @@
+            }
+        } else if (userInput.startsWith("done")) {
+            if (checkDoneCommand(userInput)) {
+                return tasks.markAsDone(userInput, tasks.getStorage().getUserInputRecords());
+            } else {
+                return "OOPS!!! I'm sorry, but I don't know what that means :-(\n";
+            }
+        } else if (userInput.equals("deleteAll")) {
+            return tasks.deleteAll(tasks.getStorage().getUserInputRecords());
+        } else if (userInput.startsWith("delete")) {
+            if (checkDeleteCommand(userInput)) {
+                return tasks.delete(userInput, tasks.getStorage().getUserInputRecords());
+            } else {
+                return "OOPS!!! I'm sorry, but I don't know what that means :-(\n";
+            }
+        } else if (userInput.startsWith("save ")) {
+            return tasks.getStorage().save(userInput);
+        } else if (userInput.startsWith("load ")) {
+            return tasks.getStorage().load(userInput);
+        } else if (userInput.startsWith("find")) {
+            return tasks.search(userInput, tasks.getStorage().getUserInputRecords());
+        } else if (userInput.equals("help")) {
+            StringBuilder builder = new StringBuilder();
+            builder.append("todo <description>\n");
+            builder.append("deadline <description> /by <time in format yyyy-mm-dd>\n");
+            builder.append("event <description> /at <time in format yyyy-mm-dd>\n");
+            builder.append("save <directory>\n");
+            builder.append("load <directory>\n");
+            builder.append("done <number>\n");
+            builder.append("delete <number>\n");
+            builder.append("deleteAll\n");
+            builder.append("find <keyword>\n");
+            return builder.toString();
+        } else if (userInput.startsWith("todo ") || userInput.startsWith("deadline ")
+                || userInput.startsWith("event ")) {
+            return tasks.add(userInput, tasks.getStorage().getUserInputRecords());
+        } else {
+            return "OOPS!!! I'm sorry, but I don't know what that means :-(\n";
+        }
+    }
```
Maybe you can abstract this method more? 
![image](https://user-images.githubusercontent.com/54053213/131607644-db245d28-98b5-4246-8987-0595b5789830.png)

ZhaoPeiduo replied:
Thanks for pointing this out. I have abstracted the construction of help message to anothe private method in the latest commit!

marcuspeh commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,190 @@
+    /**
+     * Automatically(without any user command) loads the users inputs from ../data/record when the program is restarted.
+     */
+    public void autoLoad() {
+        try {
+            Scanner scanner = new Scanner(Paths.get("data", "record"));
+            while (scanner.hasNextLine()) {
+                String itemInfo = scanner.nextLine();
+                if (itemInfo.startsWith("[T]")) {
+                    Task task = new ToDo(itemInfo.substring(7));
+                    if (itemInfo.contains("[X]")) {
+                        task.setDone(true);
+                    }
+                    userInputRecords.add(task);
+                } else if (itemInfo.startsWith("[D]")) {
+                    Task task = new Deadline(itemInfo.substring(7, itemInfo.indexOf("(by")),
+                            convertToLocalTime(itemInfo.substring(itemInfo.indexOf("(by") + 5,
+                                    itemInfo.length() - 1)));
+                    if (itemInfo.contains("[X]")) {
+                        task.setDone(true);
+                    }
+                    userInputRecords.add(task);
+                } else if (itemInfo.startsWith("[E]")) {
+                    Task task = new Event(itemInfo.substring(7, itemInfo.indexOf("(at")),
+                            convertToLocalTime(itemInfo.substring(itemInfo.indexOf("(at") + 5,
+                                    itemInfo.length() - 1)));
+                    if (itemInfo.contains("[X]")) {
+                        task.setDone(true);
+                    }
+                    userInputRecords.add(task);
+                }
+            }
+            scanner.close();
+        } catch (IOException e) {
+            System.out.println("Saved data not found, a new data file created.\n");
+        }
+    }
```
Perhaps you can abstract or simplify this method more. It's quite hard to follow the logic behind this with so much with the nested while and if(s) loops.

ZhaoPeiduo replied:
Thanks for the suggestion! I have abstracted out the add task process to another private method in the latest commit!

marcuspeh commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+     * Two deadline objects are equal iff they are equal tasks and they have the same ddl.
+     * */
+    @Override
+    public boolean equals(Object another) {
+        if (another instanceof Deadline) {
+            Deadline anotherTask = (Deadline) another;
+            return super.equals(another) && ddl.equals(anotherTask.getDdl());
+        }
+        return false;
+    }
```
Was this equals method used anywhere?

ZhaoPeiduo replied:
Yes, it is used in the JUnit test ParserTasklistTest to compare whether the expected and actual list of tasks are equal.

marcuspeh commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,45 @@
+ * Represents a specific type of Task with an additional time field.
+ */
+public class Event extends Task {
+    private LocalDate time;
```
Since this is saving the date instead of time, maybe you can rename this variable to date instead.

ZhaoPeiduo replied:
Thanks for pointing this out. I have changed the variable name to date in the latest commit!

marcuspeh reviewed (commented):
Overall quite well done. I can hardly find anything to comment on and those that I mentioned is more of trivial improvements. Good job!

ZhaoPeiduo committed:
Add color for dialog boxes
Fix coding standard issues as raised in peer review

ZhaoPeiduo commented:
@marcuspeh Thanks for the helpful comments!

yunpeng1234 commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,72 @@
+import java.awt.*;
```
Should not import all as * as stated in the coding violation. Also, all classes in should in a package so the package statement is missing here.

yunpeng1234 commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,72 @@
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.*;
```
Likewise, here the use of * is not allowed.

yunpeng1234 commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,72 @@
+import javafx.scene.paint.Paint;
+import javafx.scene.shape.Circle;
+
+// This class is taken from JavaFX tutorial 4 by Jeffry Lum https://se-education.org/guides/tutorials/javaFxPart4.html
```
Redundant comment that should be deleted.

yunpeng1234 commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,60 @@
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.*;
```
Same issue as above.

yunpeng1234 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,129 @@
+        this.tasks = tasks;
+    }
+
+    /**
+     * This method checks if a String starting with delete is a delete command.
+     * A delete command is a string with the keyword delete followed by numbers and spaces only.
+     * Examples of valid user inputs:
+     * -> delete 3
+     * -> delete   4(spaces trimmed)
+     * -> delete 1 4(spaces trimmed and interpret as done 14)
+     * Examples of invalid user inputs:
+     * -> delete task
+     * -> delete 2 task
```
Description of the method should be simple. Additional comments on the code can be done in the readme.md

yunpeng1234 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     */
+    public String markAsDone(String userInput, ArrayList<Task> userInputRecords) {
+        try {
+            int itemToComplete = Integer.parseInt(userInput.replaceAll("[^0-9]", "")) - 1;
```
Variable could be renamed to sound like a number. Eg. ItemIndex

yunpeng1234 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     */
+    public String search(String userInput, ArrayList<Task> userInputRecords) {
+        String keyword = userInput.replace("find", "").trim();
+        ArrayList<Task> searchResult = new ArrayList<>();
```
Could pluralise the search result for >1 returned result

yunpeng1234 reviewed (commented):
Overall no major coding standard violation. LGTM.

Quality wise, the code is bloated with own LocalDate formatter, I suggest you use their formatter to cut down on code bloat. 
Testing of storage should be simple, no try catch while loops inside. Suggest to use stub to extract of storage for easier testing as well as more readable tests.

ZhaoPeiduo committed:
Complete following improvements based on feedback from peer review:
- Use parse() and toString() from LocalDate
- Make a stub class for StorageTest

ZhaoPeiduo commented:
@yunpeng1234 Thanks for the helpful suggestions! I have improved my code based on your suggestions

ZhaoPeiduo committed:
Complete minor improvement in GUI design

ZhaoPeiduo committed:
Fix bug in Storage class

ZhaoPeiduo committed:
Edit the code to follow the coding standard

ZhaoPeiduo committed:
Complete A-Assertions by using assertions in Storage and TaskList

Currently, Storage::addTask and TaskList::add has an empty else block which can be replaced with an assertion

In this commit, the else blocks in these two methods are deleted and the last else if block is changed to else block starting with an assertion

This is to reduce unnecessary components (in this case the else blocks) in the code.

ZhaoPeiduo committed:
Merge pull request #1 from ZhaoPeiduo/branch-A-Assertions

Complete A-Assertions by using assertions in Storage and TaskList

ZhaoPeiduo committed:
Merge branch 'branch-A-Assertions'

ZhaoPeiduo committed:
Merge branch 'master' of https://github.com/ZhaoPeiduo/ip

ZhaoPeiduo committed:
Complete A-Code-Quality
- Storage: In Storage::addTask, the expressions to create task objects were complicated, so the description and date parameters are created separately
- TaskList: TaskList::add was a long method and three private methods were created to generate different types of Task object
- TaskList: TaskList::delete and TaskList::markAsDone have similar method body, so a private method is used to abstract the common parts out

ZhaoPeiduo committed:
Merge pull request #2 from ZhaoPeiduo/branch-A-CodeQuality

Complete A-Code-Quality

ZhaoPeiduo committed:
Merge remote-tracking branch 'origin/master'

ZhaoPeiduo committed:
Complete C-Update

- Syntax: update <index> /to <Task formatted in Task/Deadine/Event>
- Update the task with the indicated index and print the new list
- Corresponding JUnit test written

ZhaoPeiduo committed:
Update README.md

ZhaoPeiduo committed:
Fix the bug when creating Deadline and Event objects

In the previous version, the program will crash if the input starts with deadline but does not have a /by or /at

The bug is fixed by deleting TaskList::add and make Task::addToDo, addDeadline and addEvent to be public

ZhaoPeiduo committed:
Merge branch 'master' of https://github.com/ZhaoPeiduo/ip

ZhaoPeiduo committed:
Complete A-BetterGUI

- Present the profile pictures in circles
- Add background color to the dialog boxes
- Add a menu bar with close and help items
- Change the display name and icon of the app

ZhaoPeiduo committed:
Finalise the features by merging the branch 'branch-C-Update'
Fix minor checkstyle issues

ZhaoPeiduo committed:
Update the user guide and add a product screenshot

ZhaoPeiduo committed:
Modify product screenshot

ZhaoPeiduo committed:
Update UserGuide based on peer review

ZhaoPeiduo committed:
Add missing list command to UG
